File: src/screens/HomeScreen.js
import { useCallback, useMemo, useState } from 'react';
import {
  View,
  Text,
  StyleSheet,
  FlatList,
  TouchableOpacity,
} from 'react-native';
import { useFocusEffect } from '@react-navigation/native';
import { MaterialCommunityIcons } from '@expo/vector-icons';
import { getWorkouts, getLastLevel, setLastLevel } from '../storage';
import { getExercise } from '../data/exercises';
import { totalPoints, rankFor, levelOf, maxLevel, ranks } from '../data/ranks';
import { useTheme } from '../ThemeContext';
import { useDiscipline } from '../DisciplineContext';
import { useLevelUp } from '../LevelUpContext';
import DisciplineSwitch from '../components/DisciplineSwitch';
import Medal from '../components/Medal';
import FadeInView from '../components/FadeInView';
import * as Haptics from 'expo-haptics';
import { spacing, radius, glowShadow } from '../theme';

function formatDate(iso) {
  const d = new Date(iso);
  return d.toLocaleDateString('it-IT', {
    day: '2-digit',
    month: 'short',
    year: 'numeric',
  });
}

function setSummary(s) {
  const ex = getExercise(s.exerciseId);
  if (!ex) return '';
  if (ex.unit === 'weight') return `${ex.name} ${s.weight || 0}kg×${s.reps || 0}`;
  if (ex.unit === 'sec') return `${ex.name} ${s.sec || 0}s`;
  return `${ex.name} ×${s.reps || 0}`;
}

export default function HomeScreen({ navigation }) {
  const { theme, glowActive } = useTheme();
  const { discipline } = useDiscipline();
  const { celebrate } = useLevelUp();
  const styles = useMemo(() => makeStyles(theme.colors), [theme]);
  const [workouts, setWorkouts] = useState([]);

  useFocusEffect(
    useCallback(() => {
      let active = true;
      getWorkouts(discipline).then(async (w) => {
        if (!active) return;
        setWorkouts(w);
        // Rilevamento level up.
        const pts = totalPoints(w);
        const lvl = levelOf(rankFor(pts).current.id);
        const prev = await getLastLevel(discipline);
        if (prev !== null && lvl > prev) {
          celebrate(ranks[lvl - 1], lvl);
        }
        if (prev === null || lvl !== prev) {
          await setLastLevel(discipline, lvl);
        }
      });
      return () => {
        active = false;
      };
    }, [discipline])
  );

  const totalWorkouts = workouts.length;
  const totalSets = workouts.reduce((acc, w) => acc + w.sets.length, 0);
  const points = totalPoints(workouts);
  const { current, next, progress } = rankFor(points);

  const renderItem = ({ item, index }) => (
    <FadeInView delay={Math.min(index, 8) * 55}>
      <TouchableOpacity
        style={styles.card}
        onPress={() => navigation.navigate('WorkoutDetail', { id: item.id })}
      >
        <Text style={styles.cardDate}>{formatDate(item.date)}</Text>
        <Text style={styles.cardExercises} numberOfLines={1}>
          {item.sets.length} set • {item.sets.map(setSummary).slice(0, 2).join(', ')}
        </Text>
      </TouchableOpacity>
    </FadeInView>
  );

  return (
    <View style={styles.container}>
      <DisciplineSwitch />

      <FadeInView>
      <TouchableOpacity
        style={styles.rankCard}
        onPress={() => navigation.navigate('Medaglie')}
      >
        <Medal rank={current} progress={progress} size={64} />
        <View style={{ flex: 1 }}>
          <Text style={styles.rankLevel}>
            Livello {levelOf(current.id)}/{maxLevel}
          </Text>
          <Text style={[styles.rankName, { color: current.color }]}>{current.name}</Text>
          <Text style={styles.rankPts}>
            {points} pt{next ? ` • ${next.min - points} al ${next.name}` : ' • max!'}
          </Text>
        </View>
        <Text style={styles.rankChevron}>›</Text>
      </TouchableOpacity>
      </FadeInView>

      <FadeInView delay={120} style={styles.statsRow}>
        <View style={styles.statBox}>
          <Text style={styles.statNumber}>{totalWorkouts}</Text>
          <Text style={styles.statLabel}>Allenamenti</Text>
        </View>
        <View style={styles.statBox}>
          <Text style={styles.statNumber}>{totalSets}</Text>
          <Text style={styles.statLabel}>Set totali</Text>
        </View>
        <TouchableOpacity
          style={styles.statBox}
          onPress={() => navigation.navigate('Timer', {})}
        >
          <MaterialCommunityIcons name="timer-outline" size={26} color={theme.colors.primary} />
          <Text style={styles.statLabel}>Timer</Text>
        </TouchableOpacity>
      </FadeInView>

      <FlatList
        data={workouts}
        keyExtractor={(item) => item.id}
        renderItem={renderItem}
        contentContainerStyle={styles.list}
        ListEmptyComponent={
          <Text style={styles.empty}>
            Nessun allenamento registrato.{'\n'}Tocca + per iniziare!
          </Text>
        }
      />

      <TouchableOpacity
        style={[styles.fab, glowActive && glowShadow(theme.glow, 16)]}
        onPress={() => {
          Haptics.impactAsync(Haptics.ImpactFeedbackStyle.Medium);
          navigation.navigate('NewWorkout');
        }}
      >
        <Text style={styles.fabText}>+</Text>
      </TouchableOpacity>
    </View>
  );
}

const makeStyles = (c) =>
  StyleSheet.create({
    container: { flex: 1, backgroundColor: c.bg },
    rankCard: {
      flexDirection: 'row',
      alignItems: 'center',
      gap: spacing.md,
      backgroundColor: c.card,
      borderRadius: radius.md,
      padding: spacing.md,
      margin: spacing.md,
      marginBottom: 0,
      borderWidth: 1,
      borderColor: c.border,
    },
    rankLevel: { color: c.textMuted, fontSize: 12, fontWeight: '700' },
    rankName: { color: c.text, fontSize: 17, fontWeight: '800', marginTop: 1 },
    rankPts: { color: c.textMuted, fontSize: 12, marginTop: 2 },
    rankChevron: { color: c.textMuted, fontSize: 28, fontWeight: '300' },
    statsRow: { flexDirection: 'row', padding: spacing.md, gap: spacing.md },
    statBox: {
      flex: 1,
      backgroundColor: c.card,
      borderRadius: radius.md,
      padding: spacing.md,
      alignItems: 'center',
      borderWidth: 1,
      borderColor: c.border,
    },
    statNumber: { color: c.primary, fontSize: 28, fontWeight: '700' },
    statIcon: { fontSize: 26 },
    statLabel: { color: c.textMuted, fontSize: 13, marginTop: spacing.xs },
    list: { paddingHorizontal: spacing.md, paddingBottom: 90, gap: spacing.sm },
    card: {
      backgroundColor: c.card,
      borderRadius: radius.md,
      padding: spacing.md,
      borderWidth: 1,
      borderColor: c.border,
      marginBottom: spacing.sm,
    },
    cardDate: { color: c.text, fontSize: 16, fontWeight: '600' },
    cardExercises: { color: c.textMuted, marginTop: spacing.xs },
    empty: {
      color: c.textMuted,
      textAlign: 'center',
      marginTop: spacing.xl,
      lineHeight: 22,
    },
    fab: {
      position: 'absolute',
      right: spacing.lg,
      bottom: spacing.lg,
      width: 60,
      height: 60,
      borderRadius: 30,
      backgroundColor: c.primary,
      alignItems: 'center',
      justifyContent: 'center',
      elevation: 4,
    },
    fabText: { color: c.bg, fontSize: 32, fontWeight: '700', marginTop: -2 },
  });

File: lib/screens/home_screen.dart
import 'package:flutter/material.dart';
import 'package:flutter_animate/flutter_animate.dart';
import 'package:provider/provider.dart';

import '../state/theme_provider.dart';
import '../state/discipline_provider.dart';
import '../state/workout_provider.dart';
import '../state/levelup_provider.dart';
import '../state/locale_provider.dart';
import '../services/storage_service.dart';
import '../services/feedback_service.dart';
import '../data/ranks.dart';
import '../theme/app_theme.dart';
import '../utils/format.dart';
import '../widgets/discipline_switch.dart';
import '../widgets/medal.dart';
import '../widgets/animated_number.dart';
import '../widgets/ui_kit.dart';
import '../widgets/design_system.dart';
import 'new_workout_screen.dart';
import 'workout_detail_screen.dart';
import 'timer_screen.dart';
import 'store_screen.dart';

class HomeScreen extends StatefulWidget {
  const HomeScreen({super.key});

  @override
  State<HomeScreen> createState() => _HomeScreenState();
}

class _HomeScreenState extends State<HomeScreen> {
  @override
  void didChangeDependencies() {
    super.didChangeDependencies();
    WidgetsBinding.instance.addPostFrameCallback((_) => _checkLevelUp());
  }

  Future<void> _checkLevelUp() async {
    if (!mounted) return;
    final dp = context.read<DisciplineProvider>();
    final wp = context.read<WorkoutProvider>();
    final pts = totalPoints(wp.forDiscipline(dp.discipline));
    final lvl = levelOf(rankFor(pts).current.id);
    final last = await StorageService.getLastLevel(dp.discipline);
    if (last != null && lvl > last && mounted) {
      context.read<LevelUpProvider>().celebrate(ranks[lvl - 1], lvl);
    }
    if (last == null || lvl != last) {
      await StorageService.setLastLevel(dp.discipline, lvl);
    }
  }

  @override
  Widget build(BuildContext context) {
    final theme = context.watch<ThemeProvider>();
    final c = theme.colors;
    final t = context.watch<LocaleProvider>().t;
    final dp = context.watch<DisciplineProvider>();
    final wp = context.watch<WorkoutProvider>();
    final workouts = wp.forDiscipline(dp.discipline);
    final pts = totalPoints(workouts);
    final info = rankFor(pts);
    final totalSets = workouts.fold<int>(0, (a, w) => a + w.sets.length);

    return Scaffold(
      body: GradientBackground(
        child: _body(
          context,
          theme,
          c,
          t,
          dp,
          wp,
          workouts,
          pts,
          info,
          totalSets,
        ),
      ),
    );
  }

  Widget _body(
    BuildContext context,
    ThemeProvider theme,
    AppColors c,
    dynamic t,
    DisciplineProvider dp,
    WorkoutProvider wp,
    List workouts,
    int pts,
    RankInfo info,
    int totalSets,
  ) {
    return Scaffold(
      backgroundColor: Colors.transparent,
      appBar: AppBar(
        backgroundColor: Colors.transparent,
        title: ShaderMask(
          shaderCallback: (r) =>
              LinearGradient(colors: [c.primary, c.text]).createShader(r),
          child: const Text(
            'CaliStrack',
            style: TextStyle(
              color: Colors.white,
              fontWeight: FontWeight.w900,
              letterSpacing: 0.5,
            ),
          ),
        ),
        actions: [
          PopupMenuButton<String>(
            icon: const Icon(Icons.more_vert),
            onSelected: (v) {
              FeedbackService.selection();
              if (v == 'store') {
                Navigator.push(
                  context,
                  MaterialPageRoute(builder: (_) => const StoreScreen()),
                );
              } else if (v == 'timer') {
                Navigator.push(
                  context,
                  MaterialPageRoute(builder: (_) => const TimerScreen()),
                );
              }
            },
            itemBuilder: (_) => [
              PopupMenuItem(value: 'store', child: Text(t('neon_store'))),
              PopupMenuItem(value: 'timer', child: Text(t('timer'))),
            ],
          ),
        ],
      ),
      floatingActionButton:
          FloatingActionButton.extended(
            backgroundColor: c.primary,
            foregroundColor: theme.skin.isDark ? Colors.black : Colors.white,
            onPressed: () {
              FeedbackService.onTap();
              Navigator.push(
                context,
                MaterialPageRoute(builder: (_) => const NewWorkoutScreen()),
              );
            },
            icon: const Icon(Icons.add, size: 26),
            label: Text(
              t('new_workout'),
              style: const TextStyle(fontWeight: FontWeight.w700),
            ),
          ).animate().scale(
            duration: 400.ms,
            curve: Curves.easeOutBack,
            begin: const Offset(0.6, 0.6),
          ),
      body: ListView(
        padding: const EdgeInsets.all(Spacing.md),
        children: [
          const DisciplineSwitch(),
          const SizedBox(height: Spacing.sm),
          _RankCard(info: info, points: pts)
              .animate()
              .fadeIn(duration: 400.ms)
              .slideY(begin: 0.1, curve: Curves.easeOut),
          const SizedBox(height: Spacing.md),
          Row(
                children: [
                  Expanded(
                    child: _StatBox(
                      label: t('workouts'),
                      animatedValue: workouts.length,
                      icon: Icons.event_available,
                    ),
                  ),
                  const SizedBox(width: Spacing.sm),
                  Expanded(
                    child: _StatBox(
                      label: t('total_sets'),
                      animatedValue: totalSets,
                      icon: Icons.repeat,
                    ),
                  ),
                  const SizedBox(width: Spacing.sm),
                  Expanded(
                    child: PressableScale(
                      onTap: () => Navigator.push(
                        context,
                        MaterialPageRoute(builder: (_) => const TimerScreen()),
                      ),
                      child: _StatBox(
                        label: t('timer'),
                        icon: Icons.timer_outlined,
                      ),
                    ),
                  ),
                ],
              )
              .animate()
              .fadeIn(delay: 100.ms, duration: 400.ms)
              .slideY(begin: 0.1),
          const SizedBox(height: Spacing.md),
          if (workouts.isEmpty)
            Padding(
              padding: const EdgeInsets.symmetric(vertical: 40),
              child: Column(
                children: [
                  Icon(Icons.fitness_center, size: 56, color: c.textMuted)
                      .animate(onPlay: (c) => c.repeat(reverse: true))
                      .scaleXY(
                        end: 1.1,
                        duration: 1200.ms,
                        curve: Curves.easeInOut,
                      ),
                  const SizedBox(height: Spacing.md),
                  Text(
                    t('no_workouts'),
                    textAlign: TextAlign.center,
                    style: TextStyle(color: c.textMuted),
                  ),
                ],
              ),
            )
          else
            ResponsiveWrap(
              minTileWidth: 340,
              maxColumns: 2,
              children: workouts.asMap().entries.map((entry) {
                final i = entry.key;
                final w = entry.value;
                return GlowCard(
                      padding: EdgeInsets.zero,
                      radius: Radii.md,
                      onTap: () => Navigator.push(
                        context,
                        MaterialPageRoute(
                          builder: (_) => WorkoutDetailScreen(id: w.id),
                        ),
                      ),
                      child: Padding(
                        padding: const EdgeInsets.symmetric(
                          horizontal: Spacing.md,
                          vertical: 4,
                        ),
                        child: ListTile(
                          contentPadding: EdgeInsets.zero,
                          leading: CircleAvatar(
                            backgroundColor: c.primary.withValues(alpha: 0.15),
                            child: Icon(
                              dp.discipline == 'gym'
                                  ? Icons.fitness_center
                                  : Icons.sports_gymnastics,
                              color: c.primary,
                              size: 20,
                            ),
                          ),
                          title: Text(
                            formatDate(w.date),
                            style: const TextStyle(fontWeight: FontWeight.w700),
                          ),
                          subtitle: Text(
                            workoutSummary(w),
                            style: TextStyle(color: c.textMuted),
                          ),
                          trailing: Icon(
                            Icons.chevron_right,
                            color: c.textMuted,
                          ),
                        ),
                      ),
                    )
                    .animate()
                    .fadeIn(delay: (60 * i).ms, duration: 350.ms)
                    .slideX(begin: 0.08, curve: Curves.easeOut);
              }).toList(),
            ),
          const SizedBox(height: 90),
        ],
      ),
    );
  }
}

class _RankCard extends StatelessWidget {
  final RankInfo info;
  final int points;
  const _RankCard({required this.info, required this.points});

  @override
  Widget build(BuildContext context) {
    final theme = context.watch<ThemeProvider>();
    final c = theme.colors;
    final t = context.watch<LocaleProvider>().t;
    final diff = info.next != null ? info.next!.min - points : 0;
    return GlowCard(
      glow: true,
      borderColor: info.current.color.withValues(alpha: 0.6),
      child: Row(
        children: [
          Hero(
            tag: 'rank-medal',
            child: Medal(rank: info.current, size: 68, progress: info.progress),
          ),
          const SizedBox(width: Spacing.md),
          Expanded(
            child: Column(
              crossAxisAlignment: CrossAxisAlignment.start,
              children: [
                Text(
                  '${t('level')} ${levelOf(info.current.id)}/$maxLevel',
                  style: TextStyle(color: c.textMuted, fontSize: 12),
                ),
                Text(
                  info.current.name,
                  style: TextStyle(
                    color: info.current.color,
                    fontSize: 22,
                    fontWeight: FontWeight.w800,
                  ),
                ),
                Row(
                  children: [
                    AnimatedNumber(
                      value: points,
                      style: TextStyle(
                        color: c.text,
                        fontSize: 13,
                        fontWeight: FontWeight.w700,
                      ),
                      format: (v) => '${fmtCompact(v.round())} pt',
                    ),
                    Flexible(
                      child: Text(
                        info.next != null
                            ? ' • ${fmtCompact(diff)} ${t.p('to_next', {'name': info.next!.name})}'
                            : ' • ${t('max')}',
                        maxLines: 1,
                        overflow: TextOverflow.ellipsis,
                        style: TextStyle(color: c.textMuted, fontSize: 12),
                      ),
                    ),
                  ],
                ),
                const SizedBox(height: 6),
                ClipRRect(
                  borderRadius: BorderRadius.circular(4),
                  child: TweenAnimationBuilder<double>(
                    tween: Tween(begin: 0, end: info.progress),
                    duration: const Duration(milliseconds: 900),
                    curve: Curves.easeOutCubic,
                    builder: (_, v, _) => LinearProgressIndicator(
                      value: v,
                      minHeight: 6,
                      backgroundColor: c.cardAlt,
                      valueColor: AlwaysStoppedAnimation(info.current.color),
                    ),
                  ),
                ),
              ],
            ),
          ),
        ],
      ),
    );
  }
}

class _StatBox extends StatelessWidget {
  final String label;
  final int? animatedValue;
  final IconData? icon;
  const _StatBox({required this.label, this.animatedValue, this.icon});

  @override
  Widget build(BuildContext context) {
    final c = context.watch<ThemeProvider>().colors;
    return Container(
      padding: const EdgeInsets.symmetric(vertical: Spacing.md),
      decoration: BoxDecoration(
        gradient: LinearGradient(
          begin: Alignment.topCenter,
          end: Alignment.bottomCenter,
          colors: [c.cardAlt, c.card],
        ),
        borderRadius: BorderRadius.circular(Radii.md),
        border: Border.all(color: c.border),
      ),
      child: Column(
        children: [
          if (animatedValue != null)
            AnimatedNumber(
              value: animatedValue!,
              style: TextStyle(
                color: c.primary,
                fontSize: 24,
                fontWeight: FontWeight.w800,
              ),
            )
          else if (icon != null)
            Icon(icon, color: c.primary, size: 26),
          const SizedBox(height: 4),
          Text(
            label,
            style: TextStyle(color: c.textMuted, fontSize: 11),
            textAlign: TextAlign.center,
          ),
        ],
      ),
    );
  }
}
